File: src/SmartMeter.cpp

/*
 *
 * Copyright (c) 2017 Jan Fredrik Leversund
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

#include <SmartMeter.h>


SmartMeter smartmeter;

// Callback notifying us of the need to save config.
void saveConfigCallback ()
{
  Serial.println(F("Flagging config for save."));
  smartmeter.shouldSaveConfig = true;
}

void SmartMeter::setup()
{
	Serial.begin(115200);
	Serial.println(F("Starting up."));

	pinMode(BTN, INPUT_PULLUP);
	pinMode(LDR, INPUT);

	blinker.setPin(LED);
	ldr.delay = DEBOUNCE_TIME;

	readConfig();
	setupWifi();



	IPAddress addr(inet_addr(mqttServer));
	uint16_t port = atoi(mqttPort);


	Serial.print("MQTT address: ");
	addr.printTo(Serial);
	Serial.print(":");
	Serial.println(port);

	psClient.setClient(wifiClient);
	psClient.setServer(addr, port);

	snprintf(topic, MQTT_TOPIC_MAX_LEN, MQTT_TOPIC_Wh, meterId);
	Serial.print(F("Publishing to topic "));
	Serial.println(topic);
}

void SmartMeter::loop()
{
	blinker.check();
	ldr.update(digitalRead(LDR));

	if (ldr.changed && ldr.state)
	{
		counter++;
		Serial.println(counter);
		blinker.setOnForTime(10);
	}

	// Button is pressed, let's restart.
	if (!digitalRead(BTN))
	{
		Serial.println(F("Button is pressed, restarting."));
		ESP.restart();
	}

	psClient.loop();
	publish();

}


void SmartMeter::setupWifi()
{
    WiFiManagerParameter customMQTTServer("server", "mqtt server", mqttServer, CFG_MQTT_SERVER_LEN);
    WiFiManagerParameter customMQTTPort("port", "mqtt port", mqttPort, CFG_MQTT_PORT_LEN);
    WiFiManagerParameter customMeterId("id", "meter id", meterId, CFG_METER_ID_LEN);

    wifiManager.setSaveConfigCallback(saveConfigCallback);

    wifiManager.addParameter(&customMQTTServer);
    wifiManager.addParameter(&customMQTTPort);
    wifiManager.addParameter(&customMeterId);

    wifiManager.setConnectTimeout(60);

	// Indicate that we're in setup/connect mode.
	analogWrite(LED, 50);
	if (!digitalRead(BTN))
	{
		Serial.println(F("Entering configuration mode by force."));
		wifiManager.startConfigPortal(AP_NAME);
	}
	else
	{
		if (!wifiManager.autoConnect(AP_NAME))
		{
			Serial.println(F("Failed to connect and hit timeout."));
			delay(3000);
			ESP.restart();
			delay(5000);
		}
	}
	analogWrite(LED, 0);

    // If you get here you have connected to the WiFi
    Serial.println(F("Successfully connected."));

    // Read updated parameters
    strcpy(mqttServer, customMQTTServer.getValue());
    strcpy(mqttPort, customMQTTPort.getValue());
    strcpy(meterId, customMeterId.getValue());

    // Save the custom parameters to FS
    if (shouldSaveConfig)
    {
		Serial.println(F("Saving config."));
		DynamicJsonBuffer jsonBuffer;
		JsonObject &json = jsonBuffer.createObject();
		json["mqtt_server"] = mqttServer;
		json["mqtt_port"] = mqttPort;
		json["meter_id"] = meterId;

		File configFile = SPIFFS.open(CONFIG_FILE, "w");
		if (!configFile)
		{
			Serial.println(F("Failed to open config file for writing."));
		}
		else
		{
			json.printTo(Serial);
			Serial.println();
			json.printTo(configFile);
			configFile.close();
		}

    }
	Serial.print(F("Local IP: "));
	Serial.println(WiFi.localIP());
}



void SmartMeter::readConfig()
{
    Serial.print(F("Mounting file system..."));

    if (!SPIFFS.begin())
    {
		Serial.println(F(" failed!"));
		return;
	}

	Serial.println(F(" success."));
	if (!SPIFFS.exists(CONFIG_FILE))
	{
		Serial.println(F("No config file found."));
		return;
	}

	Serial.print(F("Reading config file..."));
	File configFile = SPIFFS.open(CONFIG_FILE, "r");

	if (!configFile)
	{
		Serial.println(F(" failed."));
		return;
	}

	Serial.println(F(" success!"));

	size_t size = configFile.size();
	std::unique_ptr<char[]> buf(new char[size]);

	configFile.readBytes(buf.get(), size);
	DynamicJsonBuffer jsonBuffer;
	JsonObject &json = jsonBuffer.parseObject(buf.get());

	if (!json.success())
	{
		Serial.println(F("Failed to load json config!"));
		return;
	}

	Serial.println(F("\nParsed json."));
	json.printTo(Serial);

	strcpy(mqttServer, json["mqtt_server"]);
	strcpy(mqttPort, json["mqtt_port"]);
	strcpy(meterId, json["meter_id"]);
}


void SmartMeter::publish()
{
	unsigned long now = millis();

	// Check if the client's still connected.
	if (!psClient.connected())
	{
		if (now - lastReconnect > RECONNECT_INTERVAL)
		{
			Serial.print(F("Attempting to connect to MQTT server..."));
			lastReconnect = now;
			if (psClient.connect(AP_NAME))
			{
				Serial.println(F(" success!"));
			}
			else
			{
				Serial.println(F(" failed."));
				return;
			}
		}
		else
			return;
	}

	// If we got this far, we're connected.
	if (now - lastPublish > PUBLISH_INTERVAL)
	{
		// All right, let's publish something!
		lastPublish = now;
		char payload[MQTT_PAYLOAD_MAX_LEN];
		snprintf(payload, MQTT_PAYLOAD_MAX_LEN, "%lu", counter);
		Serial.print(F("Publishing: "));
		Serial.println(payload);		
		psClient.publish(topic, payload, true);
	}

}

File: extras/smartmeter.pl
#!/usr/bin/perl -w

#
# Copyright (c) 2017 Jan Fredrik Leversund
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
# 

use POSIX "setsid";
use Net::MQTT::Simple;
use Time::HiRes qw(time);
use DBI;
use File::Path qw(make_path);

use constant MQTT_SERVER => "autolycus.radical.org";
use constant SMARTMETER_HOME => "$ENV{HOME}/.smartmeter";
use constant DATABASE_FILE => "smartmeter.db";
use constant LOG_FILE => "smartmeter.log";
use constant VOLTAGE => 230; # I happen to know that this is the voltage in my house.

sub daemonize;
sub create_tables;

# Globals
my $last_count = undef;
my $last_time = undef;


make_path(SMARTMETER_HOME);

-d SMARTMETER_HOME and -w SMARTMETER_HOME or
	die "Unable to access smartmeter data dir '" . SMARTMETER_HOME . "'";

my $dbh = DBI->connect("dbi:SQLite:dbname=" . SMARTMETER_HOME . "/" . DATABASE_FILE , "", "")
	or die $DBI::errstr;

create_tables;

my $sth = $dbh->prepare(q{
	insert into
		meter_reading
	(
		meter_id,
		read_time,
		read_period,
		delta_count,
		watts,
		ampere
	)
	values
	(
		?,
		?,
		?,
		?,
		?,
		?
	)
}) or die $DBI::errstr;

my $mqtt = Net::MQTT::Simple->new(MQTT_SERVER) ;

print "Everything seems okay; daemonizing...\n";
daemonize;

$mqtt->run(
	"smartmeter/+/Wh" => sub {
		my ($time, $topic, $count) = (time, @_);

		unless (defined $last_count)
		{
			# The first time we always seem to get a count of 0, so we'll ignore it.
			if ($count)
			{
				# The second time around we get a real reading from the smartmeter, so we save the count and value
				$last_count = $count;
				$last_time = $time;
			}
			return;
		}

		# Grab the meter id.
		$topic =~ /smartmeter\/(\d+)\//gio;
		my $id = $1;

		# Calculate the delta count and time difference between this and the previous reading.
		my $dc = $count - $last_count;
		my $dt = $time - $last_time;

		# Keep the current readings around until the next time.
		$last_count = $count;
		$last_time = $time;

		# A negative delta count probably means rollover or smartmeter reset. We'll ignore this reading.
		return if ($dc < 0);

		# Calculate the average wattage per second over the previous period.
		my $watts = sprintf("%.f", 3600 * $dc / $dt); 
		my $ampere = sprintf("%.1f", (3600 * $dc / $dt) / VOLTAGE);

#		print '#' x 40 . "\n";
#		print "Topic:            $topic\n";
#		print "Count:            $count\n";
#		print "Period count:     $dc\n";
#		print "Measuring period: $dt\n";
#		print "Watts:            $watts\n";
#		print "Ampere:           $ampere\n";
#		print "Meter ID:         $id\n\n";

		$mqtt->publish("smartmeter/$id/W" => $watts);
		$mqtt->publish("smartmeter/$id/A" => $ampere);

		$sth->execute($id, $time, $dt, $dc, $watts, $ampere)
			or die $DBI::errstr;
	}
);


sub create_tables
{
	$dbh->do(q{
		create table if not exists
			meter_reading
		(
				meter_reading_id	integer primary key asc,
				meter_id			integer not null,
				read_time			integer not null,
				read_period			real not null,
				delta_count			integer not null,
				watts				integer not null,
				ampere				real not null
		);
		
		create index if not exists read_time_idx on meter_reading (read_time);
	}) or die $DBI::errstr;
}


# Straight from the perlipc doc.

sub daemonize
{
	chdir("/")                  || die "can't chdir to /: $!";
	open(STDIN,  "< /dev/null") || die "can't read /dev/null: $!";
	open(STDOUT, "> /dev/null") || die "can't write to /dev/null: $!";
	defined(my $pid = fork())   || die "can't fork: $!";
	exit if $pid;               # non-zero now means I am the parent
	(setsid() != -1)            || die "Can't start a new session: $!";
	open(STDERR, ">&STDOUT")    || die "can't dup stdout: $!";
}

